### DISCUSSION TITLE
SAFE procedure phase 1

### DISCUSSION THREAD

**Discussion by RiannevanderFeen** [asreview/asreview#1853] [Q&A] [answered]
Hello all, 

I am hoping to use the phases of the SAFE procedure for our scoping review. 

1. My first question is about phase 1: how random = random? I was hoping to use a duplicates library from ENDNOTE from a dataset I already had collected using the search strings I am planning on using in the final dataset (but more filters). There are 130 records with potentially relevant records. So, not so random! Could I use this set as a 'random dataset' for phase 1. If not, 1). how do you recommend I collect a random set in a database and 2.) at least how many records should it hold? This is very unclear to me... Thanks for any advice!

**Comment by FelixWdm** [comment, marked as answer]
Hi @RiannevanderFeen , during the ... phase when starting your ASReview project, you can use the 'Random' option to do two things: find prior knowledge and estimate the ratio relevant/irrelevant by screening randomly:
![afbeelding](https://github.com/user-attachments/assets/8dae42bc-0912-421d-a3a7-531b15b79018)
How many records you screen randomly is up to you. I hope this helps.

**Reply by RiannevanderFeen** [reply]
Hi @FelixWdm,

Thank you for your reply! So the SAFEprocedure refers to this 'random' option in the ASReview software program?! (And not to a 'randomly selected dataset from a database')... I see. 

Greetings,
Rianne

**Reply by FelixWdm** [reply]
hi @RiannevanderFeen, this is just a workable solution for most use cases. The SAFE procedure is generic, anyform of random screening will work, but note that most database don't sort the results randomly, so active randomisation is needed. This is just one way to that.

**Comment by RiannevanderFeen** [comment]
Some realizations so far: 
1. the random mode (when adding prior knowledge) worked well in doing the phase 1 random screening at 1%. However, it was somewhat confusing that it kept saying that I had enough prior knowledge at intervals of 5 records. (I had planned to go up to 273, which is 1%). 
2. Only after completing phase 1 and having warmed up the model with the 276 records, I came to realize that this prior knowledge is automatically added tot the total list. Making the 10% requirement of the SAFE procedure a lessor heap 👍

**Reply by Rensvandeschoot** [reply]
Great observations! and keep sharing your experiences, because others can learn from your hard work!

We recently released a way to sample random records via [ASReview-Datatools](https://github.com/asreview/asreview-datatools?tab=readme-ov-file#sample).

**Reply by RiannevanderFeen** [reply]
Would this be an extra step in de SAFE procedure phases? And in which phase would this step best be implemented? Phase 4 (validation)?

**Reply by RiannevanderFeen** [reply]
@Rensvandeschoot, I would like to perform a random sample check, but the link to the new datatool doesn't help me install/open/use it. It says to use: asreview data sample command. I have no idea what that means. 

Do I write this in my command prompt? Just like that? I am clueless actually ;-)

**Reply by Rensvandeschoot** [reply]
No problem; we're here to help you!

Have you followed the [installation instructions](https://github.com/asreview/asreview-datatools?tab=readme-ov-file#installation) for ASRevievew-Datatools?

Then, indeed, you need to copy this into your command prompt:

```
asreview data sample input_dataset.xlsx output_dataset.xlsx 50
```

and replace `input_dataset.xlsx` with your dataset name, and replace the `50` if you want a different amount of records. The result should be that the 50 oldest and 50 newest records from input_dataset.xlsx are selected, plus 50 randomly chosen records (without overlap from the old and new partitions!). The resulting 150 records are written to `output_dataset.xlsx`.

**Reply by RiannevanderFeen** [reply]
Am trying it now. Installation seems to have worked. Now trying the next step. It says: 'asreview' is not recognized as an internal or external command,
operable program or batch file.

Should I put Python -m in front (which I need to do anyway eachtime I am starting up asreview lab)

**Reply by RiannevanderFeen** [reply]
Did that. It says: 

File "C:\Users\riann\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\asreview\data\base.py", line 67, in load_data
    raise FileNotFoundError(f"File, URL, or dataset does not exist: '{name}'")
FileNotFoundError: File, URL, or dataset does not exist: 'input_dataset.xlsx'

**Reply by RiannevanderFeen** [reply]
o, wait, need to replace with dataset name...

**Reply by RiannevanderFeen** [reply]
ok, so, how will I find out what my dataset name is?

**Reply by FelixWdm** [reply]
dataset was part of the filename, just replace the entire `"input_dataset.xlsx"` with your filename. To get the full file + location Shift+RightMousebutton -> Copy as path

**Reply by FelixWdm** [reply]
If you've already finsihed screening a random set for this project via ASReview Lab, this part is optional I would say.

**Reply by RiannevanderFeen** [reply]
yes, though a check on concept drift would be could to do at this stage, it seems.

However, to get the full file + location, it seem I should export the project at this point in the review? Should this be as dataset (as aposed to project)? And then choose excel? As it seems xlsx is an Excel extention?

**Reply by RiannevanderFeen** [reply]
It seemed to have worked while keeping the path as short as possible (so saved an Excel file on desktop and shortened the name). But now it says the publication year column is missing.

  File "(...)Python312\site-packages\asreviewcontrib\datatools\sample.py", line 13, in sample
    raise ValueError(f"• The input file should have a {year_column} column.")
ValueError: • The input file should have a publication_year column.  

It seems like I fixed this by changing the "year" column to "publication_year" in the Excel file. 

But then I got this message:

 FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(LABEL_NA, inplace=True)

@FelixWdm or @Rensvandeschoot:  Thanks for any advice on how to proceed from here...

**Reply by Rensvandeschoot** [reply]
Yes, the warning you received is related to a future change in pandas behavior and doesn't affect the current export process. We might consider addressing it in the future (ping @laurens88), but for now, it shouldn't prevent exporting the data.

Did you manage to export the dataset successfully?

If you're still facing issues, one thing that might help is navigating to the correct folder via the command prompt. You can do this by running: 

```
cd  C:\Users\Documents\folder_name
```
Replace the path with the name of the location where your Excel file is stored. This ensures you're working from the directory where you stored the data. This is beneficial because it reduces the chances of getting "file not found" errors when attempting to export or manipulate the data. By explicitly navigating to the folder, it simplifies subsequent commands and prevents confusion over the file's location.

**Reply by Rensvandeschoot** [reply]
The issues of `year` versus `pubication_year` will be further discussed in https://github.com/asreview/asreview-datatools/issues/50
(ping @laurens88 )

**Reply by RiannevanderFeen** [reply]
It is hard to understand what I am looking for, because I don't understand the proces. 

There was no new file downloaded on my computer, as far as I can tell. With 'the path', do you mean everything starting from: "C:..."? So I will change directory (cd) and then my file path? Should this be my initial file path (the input)? As I do not have an output file path as far as I can tell. not anything with 'output' in anycase. 

But I closed the prompt last night and do not have the history of yesterday (or do I and where can I find it?). 

I will attempt the whole procedure again and see if I get a path that I can use in the step that you are suggesting now....

**Reply by RiannevanderFeen** [reply]
@Rensvandeschoot which part of the message below is the output path (nb. the first 'C:...' is what I entered)?

C:\Users\riann>Python -m asreview data sample C:\Users\riann\Desktop\asreview_dataset_all_Student-student.xlsx output_dataset.xlsx 50
C:\Users\riann\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\asreview\io\utils.py:142: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(LABEL_NA, inplace=True)

C:\Users\riann>

Thanks!

Thanks!

**Reply by Rensvandeschoot** [reply]
First, open the command prompt. Then, navigate to the folder containing your input file by running:

```
cd C:\Users\riann\Desktop
```

This will ensure you're in the correct directory, making it easier to execute the command without specifying the entire file path.
Once you're in the correct folder, you can run:

```
Python -m asreview data sample asreview_dataset_all_Student-student.xlsx output_dataset.xlsx 50
```

In this command:

- `asreview_dataset_all_Student-student.xlsx` is the input file.
- `output_dataset.xlsx` is the output file where the sampled data will be saved.
- the output file is stored in the same folder as your input file, in your  case `C:\Users\riann\Desktop`

Let me know if it worked out!

(@VeenDuco is working on updating the [tutorials](https://asreview.github.io/asreview-academy/datatools.html) for Datatools and we will use your input!!)

**Reply by RiannevanderFeen** [reply]
Yes! This is so cool. I got an Excel file, with 150 records, entiteld: "output_dataset" on my desktop (like magic). Whohoo.

Thank you again for the timely assistance! 

Now I will create a new project in ASR L. in which I will import this radom dataset to check for concept drift. Though... I am not sure what that will look like. I will create a new discussion thread to journal about that next. :-)

**Reply by RiannevanderFeen** [reply]
As a side, I realize now I can't change directory by backspacing in the command prompt! Hahaha. (I have to get back to the riann directory to open asreview lab again, but I see how to do that now) 👍

**Reply by RiannevanderFeen** [reply]
Note: while uploading the output_dataset in ASR.L. I noticed it automatically added 5 relevant and 5 irrelevant records.

**Reply by Rensvandeschoot** [reply]
Then, there is a column with labeling decisions in your dataset :-)
